### ISSUE TITLE
Report shows '0' count for cell population and it's sub-populations

### ISSUE THREAD

**Issue by srideviks** [whitews/FlowKit#120]
**Describe the bug**
Hello team, I use FlowKit to extract gated events from fcs file and it's corresponding workspace (wsp) file. I'm interested in count of Beads, and cell populations like Lymphocytes, and it's 7 subpopulations. When I load the input files (fcs and wsp file) to a FlowKit session and apply get_gate_events(), I get events corresponding to Beads, but not for Lymphocytes and it's sub-populations. So, I checked back the report that gets created when function "analyze_samples()" function is applied to the session. It contains 0 events for Lymphocytes and it's sub-populations. I opened this wsp file in FlowJo and I see the gates for all the cell populations in the assay. I checked wsp file in xml format, information like co-ordinates for all the populations exists in there. I have 1 pair of files (fcs and wsp file) that fives me all events. Rest all files, seems to be going through this issue. Can you please help me out to overcome this issue?

**Code To Reproduce**
Code to reproduce the behavior:
```
import flowkit as fk

sample_id = 'A_TBNK.fcs' # Renamed A for simplicity
sample_grp = 'TBNK'
base_dir = "../../Data/Orig_wsp_fcs/"
sample_path = os.path.join(base_dir, sample_id)
wsp_path = os.path.join(base_dir, "A_TBNK.wsp")    # Renamed A for simplicity

fks_fj = fk.Session(sample_path)
fks_fj.import_flowjo_workspace(wsp_path)

fks_fj.analyze_samples(sample_grp)
results = fks_fj.get_gating_results(sample_grp, sample_id)
print(results.report)

bead_events = fks_fj.get_gate_events(sample_grp, sample_id, 'Beads')    # Parent: Root
lymp_events = fks_fj.get_gate_events(sample_grp, sample_id, 'Lymphocytes')    # Parent: Root
CD3minus_events = fks_fj.get_gate_events(sample_grp, sample_id, 'CD3-')    # Parent: Lymphocytes
Bcell_events = fks_fj.get_gate_events(sample_grp, sample_id, 'B cells')    # Parent: CD3-
NKcell_events = fks_fj.get_gate_events(sample_grp, sample_id, 'NK cells')    # Parent: CD3-
Tcell_events = fks_fj.get_gate_events(sample_grp, sample_id, 'T cells')    # Parent: Lymphocytes
CD4plus_events = fks_fj.get_gate_events(sample_grp, sample_id, 'CD4+')    # Parent: T cells
CD4CD8_events = fks_fj.get_gate_events(sample_grp, sample_id, 'CD4+CD8+')    # Parent: T cells
CD8plus_events = fks_fj.get_gate_events(sample_grp, sample_id, 'CD8+')    # Parent: T cells
```

**Expected behavior**
I expect to see events(~1000 rows) for all the cell populations defined above.

**Screenshots**
![image](https://user-images.githubusercontent.com/19530531/181310384-c03fd1df-2d58-4ff0-8db2-2c8894421734.png)

**Desktop (please complete the following information):**
 - OS: Windows version 10.0.19044.1766
 - Python version 3.9
 - FlowKit version 0.9.1

**Additional context**
FlowKit is really amazing and I love to work with it. Thank you for developing this package in python and making it available for all.

**Comment by whitews** [comment]
Hi Sridevi,

Any chance this sample has "custom" gates? These would be gates where the boundary was modified vs the sample group defined gates and show up in the FlowJo main gate tree with black text instead of the sample group color. If so, this could explain the discrepancy as FlowKit doesn't yet support this (working on it for the 1.0 release). There is a workaround for this case. If not, let me know...could be a bug.

Thanks,
Scott

**Comment by srideviks** [comment]
Hi Scott,

Thank you for your quick reply.
I just checked in both FlowJo and wsp file, all the cell-populations and Beads have custom gates (In wsp file under <SampleNode>, owning_group=''). Can you please let me know the workaround?

Regards,
Sridevi

**Comment by whitews** [comment]
Okay, that is likely the problem. The fix is to create a modified copy of the wsp file:

Open the wsp file in FlowJo, delete the sample group of interest. FlowJo will open a new dialog window asking if you want to move all the gates to the "All Samples" group. Click yes to the prompt. At this point all the gates will be copied to the "All Samples" group and every gate for every sample will be a "custom" gate. But, FlowKit can read these in that group.

The reason this works is complicated and related to the odd way FlowJo stores info in the XML tree. Like I said though, fixing this is the main priority for the 1.0 release, which will hopefully be finished in a couple months.

Regards,
Scott

**Comment by srideviks** [comment]
Thanks for the workaround, Scott.

I just tried it but the issue still exists for the sample(say A).
I have another sample file(say B) which contains all custom gates for the Sample Group and that works without any error in FlowKit.
So, I compared the wsp files for both sample A and B. Of course the numbers differ as they are 2 different samples, but I see one main difference in both the files for Lymphocytes gate - In sample A, it is called "EllipsoidGate", whereas in sample B it is called "PolygonGate". Is this difference in gate name causing an error in sample A, but works in sample B?

Regards,
Sridevi

**Comment by whitews** [comment]
That's odd, so the same gate name in the same file has different gate types for 2 different FCS files? Is it possible to share the wsp file and FCS file to me privately?

**Comment by srideviks** [comment]
Oh no, I wish I could share the files with you. I need permission to share. I will work on that.
Meanwhile, let me explain the scenario more in detail:
I have 2 different samples, each having fcs and wsp files. Let me name them as sampleA and sampleB.
When I run FlowKit analyze_samples() function for sampleA and sampleB, below is the screenshot of result report:
![image](https://user-images.githubusercontent.com/19530531/181368333-621ff3e1-c80e-42eb-9ad2-e99690f9e238.png)

We can see that in sampleA the lymphocyte gate is EllipsoidGate and I get 0's for lymphocytes and it's sub-populations. Whereas in sampleB, the lymphocyte gate is PolygonGate and I get correct values for lymphocytes and it's sub-populations.
I tried a some more samples with EllipsoidGate for lymphocyte, which resulted in same 0's. Is this difference in type of gate causing this issue in FLowKit?

Regards,
Sridevi

**Comment by whitews** [comment]
Might be something wrong with parsing the EllipsoidGate, but I thought I tested that. The next thing I'd test is plotting the gate in FlowKit using the `Session.plot_gate` method. Compare what it looks like to what FlowJo displays. Can you post screenshots of both FlowKit & FlowJo's plots?

**Comment by srideviks** [comment]
I will try that and send you the plots in sometime.
I tried one more approach where I replaced EllipsoidGate (along with gating coordinate) in SampleA with PolygonGate from SampleB. When I processed the modified SampleA wsp file using FlowKit, the number of cells appeared for Lymphocytes and it's sub-population. Thus, I believe the main issue (issue #120) is because of "EllipsoidGate" in wsp file.

I will share you the plots in sometime:)
Regards,
Sridevi

**Comment by srideviks** [comment]
Hi Scott,

Find the plots below for SampleA both from FlowKit(left) and FlowJo(right):
![image](https://user-images.githubusercontent.com/19530531/181375133-1afe23fe-9dce-4f0c-96e6-cbe3230537fd.png)

I just used the following code: 
``
p = fks_fj.plot_gate(sample_grp, sample_name, "Lymphocytes")
show(p)
``

Regards,
Sridevi

**Comment by whitews** [comment]
Thanks Sridevi, definitely something wrong there with the parsing of the ellipse location. I'll try and replicate it in FlowJo but it'll probably be tomorrow til I can get back to you.

**Comment by srideviks** [comment]
Thank you so much, Scott.
I really appreciate your quick responses. 
I look forward to the solution:)

Regards,
Sridevi

**Comment by whitews** [comment]
Hi Sridevi,

An update to let you know I am still working on this. I was able to replicate the issue but the fix may take some time. The problem occurs with the combo of FlowJo's ellipse gate and the use of their "biex" transform. The ellipse gate is handled by FlowJo differently than any other gate type, with the foci & vertex locations stored in the FlowJo "display" space of 256 x 256 bins. All other gate types store their locations in data space.

In the meantime, converting those ellipse gates to some similar shaped polygon gates would be your best option. I'll update this issue when I have more progress to report.

Kind regards,
Scott

**Comment by srideviks** [comment]
Thank you, Scott:)
Looking forward to the next update:)

Regards,
Sridevi

**Comment by whitews** [comment]
After a long trip down a rabbit hole, I understand what is happening here & have a fix ready. Some background for anyone interested & my own future reference:

FlowJo bins all event data in a "display" space into 256 x 256 bins. This I knew about. However, there are 2 additional complications. First, the ellipse foci and edge vertices are stored in this display space, where all other gate types store location info in the transformed data space. Second, the biex transform scales data from 0 to 4096, where all other transforms scale from 0 to 1.

Now, consider the case where a gate is drawn on a set of FCS parameters where one parameter uses the biex transform, such as  a fluorescent channel, and the other parameter uses a non-biex transform, such as a linear transform for a scatter channel. The data space axes have different scales, yet FlowJo's display space has the same scale. If the gate is an ellipse, it can only be properly defined in the display space and cannot be a true ellipse in the data space.

I also already knew that FlowJo converts all ellipse gates to approximate polygons internally when evaluating gated events. However, it now makes sense why they do this: they have to because the above scenario cannot be an actual ellipse.

FlowKit, up to this point, has always used true ellipsoids (in n-dimensions) for processing gated events, including imported FlowJo ellipse gates. This will now change such that all FlowJo ellipse gates will be converted to approximate polygon gates (with 128 vertices).

I hope to have a new release (0.9.2) out today or tomorrow with the fix.

**Comment by srideviks** [comment]
Thanks Scott.
Wonderful explanation and I truly appreciate your efforts in putting this thing together in FlowKit.

Regards,
Sridevi

**Comment by whitews** [comment]
Sridevi,

Can you clone the master branch and test that the update fixes the issue for you? If so, I'll release the new patch version so it's installable via pip.

Thanks!
Scott

**Comment by srideviks** [comment]
Hi Scott,

So sorry, I missed your message as I was completely occupied last weekend.
I will check today and get back to you.

Regards,
Sridevi

**Comment by srideviks** [comment]
Hi Scott,

I just cloned the master branch (0.9.2b0) and tested my samples with EllipsoidGate. The update works as needed. The EllipsoidGate is now read as PolygonGate in FlowKit and gets me the correct report. Thank you so much for updating this fix. 

Thank you once again:)
Regards,
Sridevi

**Comment by whitews** [comment]
Hi Sridevi, 

I was on vacation last week. Just released 0.9.2. Thanks for reporting and testing!

All the best,
Scott

**Comment by srideviks** [comment]
Thanks for the new release, Scott:)
Hope you had nice vacation.

Regards,
Sridevi
